### PR TITLE
[SPARK-17031][SQL] Add `Scanner` operator to wrap the optimized plan directly in planner

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Added `Scanner` operator to wrap the optimized plan directly in planner, it holds project lists as well as filter predicates.
Updated relative `Analyzer` and `Optimizer` rules.
## How was this patch tested?

Existing testcases.

### CODE REVIEW

**Comment by hvanhovell** [comment]
Could you elaborate on what you are trying to do here? I am missing context here. What is the advantage of doing this?

**Comment by jiangxb1987** [comment]
@hvanhovell This idea is inspired by @cloud-fan, as he stated in [comment](https://github.com/apache/spark/pull/13893#discussion_r73305098), we'd better have a wrapper node for scan, so that the planner may match the wrapper node directly instead of resolving the whole plan using `PhysicalOperator`.

**Comment by rxin** [comment]
I'm pretty confused by this as well. Is this just collapsing Filter, Project, and an arbitrary node into a single logical node?

**Comment by cloud-fan** [comment]
see discussion here: https://github.com/apache/spark/pull/13893#discussion_r73117855

Currently we collect the projects and filters on scan node at planner by `PhysicalOperator.unapply`. The `PhysicalOperator.unapply` mostly duplicates the logic from column pruning and filter push down rules in optimizer, but doesn't handle non-deterministic expressions well. By adding a wrapper node on scan, we can push down projects and filters to scan node at optimizer phase, and reuse the existing rules. Thus we can eliminate the duplicated codes and handle non-deterministic expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1636,6 +1638,30 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
 }
 
 /**
+ * Insert a [[Scanner]] operator over [[MultiInstanceRelation]], then combine [[Project]]s and
+ * [[Filter]]s with this operator, then the planner can match on the [[Scanner]] node directly.
+ */
+object InsertRelationScanner extends Rule[LogicalPlan] with PredicateHelper {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This doesn't reuse the existing `ColumnPruning` and `PushDownPredicate` rules. I'd like to add the wrapper at the very beginning, e.g. `SessionCatalog.lookUpRelation`.

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Roger that, will change it and retest.

**Comment by jiangxb1987** [comment]
@cloud-fan I've moved the `InsertRelationScanner` rule to `Analyzer`, after relations and expressions are resolved. To reuse analyze and optimize rules, I updated relative rules such as `CleanupAliases`、 `ColumnPruning`、 `PushDownPredicate`、 `InferFiltersFromConstraints`、 `ConvertToLocalRelation`、 `PropagateEmptyRelation`, I also added new rules to combine and prune `Scanner` operators. Besides, I made some change in subquery related rules and recently found they have been refactored.
Now that only a few of test cases is still failing, which should be easy to fix. But, I realized adding a wrapper node over every relation maybe not a idea that is perfect enough for the following reasons:
Firstly, scan a relation is not among basic operators in SQL language, when we declare a relation, we imply it should be scanned, so It seems semantically duplicate to declare a `Scanner` node over a relation or calling `relation.scanner()`. Besides, to add this wrapper node, we have to make a new assumption that no other operators should be inserted between `Scanner` and its corresponding relation, this brought in more complexity.
Secondly, a wrapper node should contain the output, predicates that can be used in partition pruning, and a relation to be scanned. But this may cause complex situation in some cases, for example, in `InferFiltersFromConstraints`, we have to covert expression in filters to alias name when we collect valid constraints, because output maybe alias and filters have to use child expression, this behavor is not needed in other operators.
At last, I feel adding such a operator have caused too many changes, perhaps we should make some improvement on `PhysicalOperation`, until we figure out a way comprehensively better than current method.

After all, I'm passionate to this improvement and will try my best to contribute, please correct me if I'm wrong, thank you!

**Pull request closed without merging**